src/jmatrix.jl

```diff
@@ -1,110 +1,108 @@
 
-struct JMatrix{T, N<:Integer, SGN} <: AbstractMatrix{T}
-    n::N    #size of the square matrix
-    sgn::SGN #+-1, allows to transpose,invert easily the matrix.
-    function JMatrix{T, N, SGN}(n, sgn) where {T, N<:Integer, SGN}
-        (sgn == T(1) || sgn == T(-1) ) || throw("sgn argument must be +-1")
-        new{T, N, SGN}(n,sgn)
-    end
-end
-"""
-    JMatrix(T, n, sgn)
 
-Creates an `AbstractMatrix{T}` of size n x n. The JMatrix 
-is a Block-diagonal matrix whose diagonal blocks are [0 1; -1 0].
-If n is odd, then the last block is the 1 x 1 zero block.
-sgn is +-1, set to 1 by default. sgn allows to transpose and invert 
-the JMatrix easily. If sgn=-1, the matrix is transposed.
 """
-JMatrix(T::Type, n::Integer) = JMatrix{T,typeof(n),Any}(n, T(1))
-JMatrix(T::Type, n::Integer, sgn) = JMatrix{T,typeof(n), Any}(n, T(sgn))
-
-Base.similar(J::JMatrix,::Type{T}) where {T} = JMatrix(T, J.n, J.sgn)
-Base.similar(J::JMatrix, ::Type{T}, dims::Union{Dims{1},Dims{2}}) where {T} = zeros(T, dims...)
-Base.copy(J::JMatrix{T}) where T = JMatrix(T, J.n, J.sgn)
+JMatrix{T, ±1}(n)
 
-Base.size(J::JMatrix) = (J.n, J.n)
-function Base.size(J::JMatrix, n::Integer)
-    if n == 1 || n == 2
-        return J.n
-    else 
-        return 1
+Creates an `AbstractMatrix{T}` of size `n x n`, representing a
+block-diagonal matrix whose diagonal blocks are `±[0 1; -1 0]`.
+If `n` is odd, then the last block is the `1 x 1` zero block.
+The `±1` parameter allows us to transpose and invert the matrix,
+and corresponds to an overall multiplicative sign factor.
+"""
+struct JMatrix{T<:Real, SGN} <: AbstractMatrix{T}
+    n::Int    # size of the square matrix
```

Comment:
Note that there is not much point in storing matrix sizes as anything other than `Int`.   Also, note that the value of the sign can be stored in the type parameter `SGN` itself.